### ISSUE TITLE
Can't install, claims there is already an installation

### ISSUE THREAD

**Issue by PhilterPaper** [StrawberryPerl/Perl-Dist-Strawberry#133]
Where do 5.36 and 5.38 (.msi packages) try to install? With 5.32 and before, it would install into \Strawberry, which I could rename to \Strawberry-5.xx and generate a soft link ("junction") from \Strawberry. This way, I could have multiple Strawberry Perls and switch between them at will (using \Strawberry in my PATH, etc.). I just tried 5.38 (and then 5.36), after rmdir \Strawberry, and all I get is an error popup: "Already installed version of Strawberry Perl (64-bit), cannot be upgraded; it has to be uninstalled first." What needs to be uninstalled, and how? Is there some configuration data hiding somewhere? I'm certainly not going to uninstall \Strawberry-5.32, 26, and 24 -- it is critical to have multiple versions available for testing.

moved from #94

**Comment by shawnlaffan** [comment]
Thanks for opening a new issue.  I'll add my previous response from #94 to save some clicking.  

> That part of the build system is unchanged from the 5.32 builds, for which the default is C:\strawberry. Unless something on the build machine has been baked-in? Or perhaps some registry setting is now triggering at install time?

Another possibility is a change in the WiX tool used to build the MSIs.  SP 5.36 and 5.38 used Wix 3.11.2.  I am not sure which version SP 5.32 used.

**Comment by PhilterPaper** [comment]
Could someone try installing 5.36, rename \Strawberry to \Strawberry-5.36, and then try installing 5.38? If it fails, there's something wrong (or at least, greatly changed) with the installer.

I don't know if and how Strawberry "bakes" itself in, so that a subsequent install sees it's "there". I never had a problem with it before (5.18 through 5.26, plus 5.32). I could just rename \Strawberry to \Strawberry-5.xx, and make a Junction with no problem. I never touched the Windows registry, as far as I am aware of.

**Comment by shawnlaffan** [comment]
To be honest I have not looked at what exactly it does, but I do know the relocatable version ("Zip edition") sets registry entries so assume the MSI does also.  (We have not distributed the relocatable versions for 5.36 and 5.38).  

If someone could test it as suggested above then that would be very useful.

**Comment by sisyphus** [comment]
> Could someone try installing 5.36, rename \Strawberry to \Strawberry-5.36, and then try installing 5.38? 

Yep - it's as @PhilterPaper says.
(Another incentive to instead use the portable versions.)

I wondered whether removing the C:\Strawberry directories from the PATH prior to attempting the installation of 5.38 might be all that's needed.
I think that's not likely, but I'll give it a try if I ever find instructions (that I can follow) on how to remove those directories permanently.
In the meantime, the C:\Strawberry folders can remain in my PATH. There won't be any such directories on this PC.

Cheers,
Rob

**Comment by PhilterPaper** [comment]
I just tried removing the three \Strawberry entries in the System Path, as @sisyphus suggested, and got the same failure on install, so that doesn't seem to be it. At any rate, it doesn't appear to be checking %PATH% for \Strawberry.

I'm on Win 10, but assuming it's similar on Win 11, click on the Windows logo to bring up the Start Menu, select the Settings (gear icon). When it asks what setting to search for, enter "environment" (System Properties > Advanced > Environment Variables). Select the system variables, and it will list them all, one per line. They can be deleted, changed, inserted, or moved around.

**Comment by shawnlaffan** [comment]
RegEdit can be risky when making changes but it won't hurt to see what is being set.  

Can you run it and search for "Strawberry Perl"?

**Comment by PhilterPaper** [comment]
Looking specifically for "Strawberry Perl", there are quite a few hits (even more if looking for "Strawberry\" or just "Strawberry"). Most seem to be in the Installer-related (and CurrentVersion and Uninstall) areas, with quite a few involving "Start Menu". Registry appears to hoard everything it's ever seen, including a list of every Strawberry-related directory and its contents. I don't dare to take a machete and start hacking away at this. Most of the Strawberry stuff persists even if I rmdir \Strawberry.

It would be reasonable for MSI (Microsoft Installer, I presume) to try to avoid overwriting an existing installation, but at least it should give you that option (after a warning). Could you check if you have .msi configured to flat out refuse to overwrite an existing installation?

**Comment by shawnlaffan** [comment]
Thanks for checking.  

At this point I'll need to defer to people with more experience in wrangling MSI files and/or registry settings.  Hopefully some of the watchers can chime in.

**Comment by PhilterPaper** [comment]
Google `msi configure overwrite installation` and look at the first couple of entry clusters returned. It sounds like _possibly_ you missed updating a version number or product ID when you created the .msi. Note that I do **not** actually want to _overwrite_ an existing installation -- I want to create a new one. Even though I renamed the old ones to \Strawberry-5.xx, MSI seems to think that there is still an existing Strawberry Perl installation on the system, of the same version number.

If I need to do anything besides renaming a just-installed \Strawberry directory to \Strawberry-5.xx, I'm all ears... Or, if using one of the alternate installation methods will produce the same effect as an msi install (without a bunch of messy registry-diving), I'm willing to to do that. I see a number of suggestions that MSI is best for corporate rollouts, and may not be necessary for me.

**Comment by sisyphus** [comment]
IMO, the "portable" distros are the way to go if you want multiple versions of Strawberry Perl. I have numerous such installations.
There's even berrybrew (https://github.com/dnmfarrell/berrybrew) which, I believe, enables you to manage and switch between  the different versions.
(Disclaimer: I have not actually used berrybrew.)

If you want to stick with the MSI,  I've just verified that the following procedure works for me.
Having installed the 5.36.1 MSI, create a copy of the "Strawberry" directory named (say) "Strawberry-5.36.1".
Then re-execute the 5.36.1 MSI - this will give you the option to remove "Strawberry", which is the option you take.
Then execute the 5.38.0 MSI to install 5.38.0 into the "Strawberry" location.
Then you can switch from one to the other with appropriate renaming of the folders:
```
C:\>perl -le "print $];"
5.038000

C:\>rename Strawberry Strawberry-5.38.0

C:\>rename Strawberry-5.36.1 Strawberry

C:\>perl -le "print $];"
5.036001
```
When it comes to installing 5.40.0 somewhere down the track, you can first repeat the clunky procedure of creating a copy of, and then remoiving, "Strawberry".
But I don't know what happens if it turns out that the version of perl in the "Strawberry" folder is not the version that MSI thinks it is. (Best to avoid that, methinks.)

Cheers,
Rob

**Comment by PhilterPaper** [comment]
Hmm. OK, I think the last version I installed was 5.32.1.1 64 bit (MSI), so I got another copy of that MSI. I xcopy'd \Strawberry-5.32 to \Strawberry, and ran the 5.32 MSI. I got the expected message that there was already an installation, and only had the option to "Finish" (quit) the abortive new installation. I can't find an uninstaller for Strawberry Perl: if I select from the list of installed programs in the Start Menu, there _is_ a Strawberry Perl, but the sublist doesn't include any sort of uninstall or remove. If I go to the generic Add/Remove programs, again there is a Strawberry Perl, but the confirmation dialog says it's 5.26 (which is where I currently do most of my work, and I don't want to uninstall it yet). No option to remove 5.32. Any ideas?

BTW, I create a "junction" (symbolic link) from \Strawberry-5.xx to \Strawberry, rather than renaming files. I removed the \Strawberry junction and xcopy'd 5.32 to \Strawberry, but it doesn't seem to have recognized it as the "current" install (although the `print $];` _did_ show 5.032001). Soon I will give the "portable" method a try. I still think there is a problem with the 5.36 and 5.38 MSI creations, as I never had problems before installing a new \Strawberry and renaming it.

**Comment by sisyphus** [comment]
>  I got the expected message that there was already an installation, and only had the option to "Finish" (quit) the abortive new installation.

Ok - I'm thinking that means that the MSI is seeing a Strawberry Perl version that is other than 5.32.1. (That's the scenario I was suggesting that you avoid - but I now see that you don't have to worry about that because it doesn't enable any changes.)
If the MSI believed that 5.32.1 was, in fact, installed then (IME) you would have been given the option to either repair or remove it.

I would just try running other Strawberry MSI packages until you find one that gives you that "remove" option. 
Of course, in doing that. you need to ensure that the MSI you're executing matches the version of Perl that's actually in the Strawberry folder.
I wonder if there's some way of determining which version the registry thinks is installed ?

It might require that you have to click "next" one or two times before the "remove" option appears - but when "finish" is the only option then there's obviously nothing else that you can do.

> but it doesn't seem to have recognized it as the "current" install (although the print $]; did show 5.032001). 

The fact that it showed 5.032001 doesn't mean that the MSI thinks that version 5.32.1 is installed.
All it means is that 5.32.1 is the version that the one-liner loaded.

> I still think there is a problem with the 5.36 and 5.38 MSI creations, as I never had problems before installing a new \Strawberry and renaming it.

FAIK, it is indeed quite possible that something has changed.

I'll install 5.32.1 MSI later today and see how this beautiful hack of mine goes ;-)
According to my (sometimes unreliable) recollection, the installation of multiple MSI versions was never actively supported.


Cheers,
Rob

**Comment by shawnlaffan** [comment]
It could well be that the MSI for 5.38 has some of the same values as that for 5.36.  

FWIW, I did use the same template file for both but the relevant version numbers should have been substituted in as needed, and this is for the exe and not the MSI.
https://github.com/StrawberryPerl/Perl-Dist-Strawberry/blob/master/share/perl-5.36/perlexe.rc.tt

The relevant code for the MSI is here, but I don't see anything immediate that looks odd and it has not been changed in seven years: https://github.com/StrawberryPerl/Perl-Dist-Strawberry/blob/master/lib/Perl/Dist/Strawberry/Step/OutputMSI.pm

More generally, though, the portable versions will provide a much cleaner environment for what @PhilterPaper is trying to achieve.  Each can be unzipped to its own directory and junctioned as needed.  The system then only needs to have the relevant paths in ```%PATH%```, e.g. ```C:\strawberry\perl\site\bin;C:\strawberry\perl\bin;C:\strawberry\c\bin```.

**Comment by sisyphus** [comment]
To install the 5.32.1 MSI, I first need to place a copy of my "C:\Strawberry-5.38.0" in "C:\Strawberry".
Next I need to uninstall that perl-5.38.0 that's now in C:\Strawberry.
I do that by executing (double clicking on) the 5.38.0 .msi file.
That brings up the Strawberry Perl Setup box and I click on "Next".
That brings up the option to select either "Repair" or "Remove", so I select "Remove", then click on "Remove", then click on "Finish" when that option finally appears.
The C:\Strawberry directory is then either no longer existent or is empty. (I'm not sure which, but it doesn't really matter.)
However, I still have C:\Strawberry-5.38.0 and C:\Strawberry-5.36.1.
Then I execute the 5.32.1 MSI and perl-5.32.1 gets installed into  C:\Strawberry.
```
C:\>perl -le "print $];"
5.032001
```
I then copy that C:\Strawberry directory to C:\Strawberry-5.32.1 for safe keeping.
Now I can run MSI version 5.32.1, 5.36.1 or 5.38.0 by renaming folders appropriately.

Out of curiosity, I remove that 5.32.1 installation from C:\Strawberry folder (in preparation for installing another version of Strawberry Perl by re-executing the 5.32.1 MSI , then clicking "Next", then clicking "Remove", then clicking "Remove" then clicking "Finish".
I just wanted to check the removal of 5.32.1 worked as expected, and it did.
But there's now no perl in C:\Strawberry, and I don't really want to install a fourth MSI version of perl, so I just re-install the 5.38.0 version by re-executing the 5.38.0 MSI.

Now I can again run MSI version 5.32.1, 5.36.1 or 5.38.0 by renaming folders appropriately.

If I ever want to install a fourth MSI version of Strawberry Perl I need to firstly make sure that C:\Strawberry contains perl-5.38.0,  and to secondly execute the 5.38.0 MSI in order to remove that perl-5.38.0.

I know @PhilterPaper is messing with symlinks rather than messing with renaming of folders. I don't think that should make a difference, but I have sometimes been wrong before.

I didn't perceive any difference in the way the 5.32.1, 5.36.1 and 5.38.0 Setups behaved.

The worst thing about this hack is that, if you want to add another perl, you first have to know (or be able to work out) which perl it was that was most recently installed. (For me, that's  currently perl-5.38.0.)
I agree with @shawnlaffan  that the better approach is to use the "portable" perls.
When I use Strawberry Perl, I use the portable builds exclusively.

Cheers,
Rob

**Comment by PhilterPaper** [comment]
OK, I finally got some time to install the portable 5.38.0.1-64bit version into \Strawberry-5.38, and so far, so good! Thanks for the help. It took a while to update all the packages to the current CPAN level, and the following fail to update:
* Mojolicious 9.34
* Math::Int64::die_on_overflow 0.54
* PPM::XML::Element 2.19
* Types::Serialiser::Error 1.01
* DBD::Pg 3.17.0 (I've never been able to update this one on any Perl level)

I don't know if you particularly care about any of the above, and I don't use them anyway (FYI). It's just a minor nuisance when bulk updating all packages (I have a `cpan` wrapper that ignores them). There _are_ one or two packages I _do_ use, that currently do not install: Graphics::TIFF and (optional for Graphics::TIFF) Image::Magick. I will open a ticket against Graphics::TIFF.

I will close this ticket soon, unless you want to keep it open for questions or anything.

**Comment by sisyphus** [comment]
I'm a bit curious about Math::Int64::die_on_overflow.
Being part of Math::Int64, it's actually not a standalone distro.

When I run <code>cpan -i Math::Int64::die_on_overflow</code> it simply installs Math-Int64-0.54 (which is the latest version of Math::Int64, and already included in SP-5.38.0).

I notice that die_on_overflow.pm itself doesn't actually have a version number. I wonder if that is somehow leading to confusion over whether it is up-to-date.

Cheers,
Rob

**Comment by PhilterPaper** [comment]
Yeah, a missing `$VERSION` appears to be the culprit in Math::Int64::die_on_overflow (also native_if_available), as well as PPM::XML::Element and Types::Serialiser::Error. Since I don't use them, it's easier just to list them as "do not attempt to upgrade until next release", rather than trying to persuade their owners to add `$VERSION`.

Mojolicious fails one test in t/mojo/file.t. It's not obvious what the problem is (e.g., a Windows incompatibility); perhaps someone who actually uses the package can report it and work with the owner.
